api: Rename query_map attribute of request macro to query_all and remove bound

zecakeh opened the pull request:
Allows to represent the query fields as a single struct or enum.

This was requested with a valid use case of representing the query strings as an enum at: https://matrix.to/#/!veagCdDBjKrMsOCzrq:privacytools.io/$eNLiLj5-OXZCOyGu8fVIYlCzfZbm-EAnNw6GfT-r34E?via=matrix.org&via=envs.net&via=mozilla.org

zecakeh commented:
One could argue that we don't need both `query_map` and `query_type` since they are the same, except that `query_map` checks for an `IntoIterator<Item=(String,String)>` bound.

jplatte commented:
Yeah I would say we should only have the more general one. I wonder if another name would be more intuitive, too. But everything I've come up with so far seems awkward... `query_all` and `query_flatten` both seem suboptimal.

zecakeh commented:
I agree that the name is not ideal. I got `query_full` or `query_wrapper`…

jplatte commented:
Hm, I guess `query_all` and `query_full` are least likely to be interpreted to mean sth. else? I'd be okay with either given we have nothing better.

zecakeh committed:
api: Rename request macro attribute query_map to query_all

Remove its IntoIterator bound to allow to represent
the query fields as a single struct or enum.

zecakeh commented:
Alright, I chose `query_all`

toger5 commented:
I might be a bit too late.
What I would think could also work and match the behaviour of: `#[ruma_api(body)]` 
is to rename `#[ruma_api(query)]` to `#[ruma_api(query_parameter)]` and make `#[ruma_api(query)]` behave like `query_full`

But I guess this is too much of a breaking change.

`query_object` would also make sense imo

jplatte commented on crates/ruma-common/src/api.rs:
```diff
@@ -122,9 +122,10 @@ macro_rules! metadata {
-///   type that implements `IntoIterator<Item = (String, String)>` (e.g. `HashMap<String,
-///   String>`, can be used for cases where an endpoint supports arbitrary query parameters.
+/// * `#[ruma_api(query_all)]`: Instead of individual query fields, one query_all field, of any
+///   type that can be serialized by [serde_html_form], can be used for cases where multiple
```
```suggestion
///   type that can be (de)serialized by [serde_html_form], can be used for cases where multiple
```

jplatte commented:
`query_object` also seems better than `query_type`, but I think it's less self-explanatory than `query_all`.

toger5 commented:
Thank you two for getting this merged so quickly!
